runtime/obj.c

```diff
@@ -85,33 +85,48 @@ CAMLprim value caml_obj_block(value tag, value size)
   sz = Long_val(size);
   tg = Long_val(tag);
 
-  /* It is difficult to correctly use custom objects allocated through
-   * [Obj.new_block]. The first field of a custom object must contain a valid
-   * pointer to a block of custom operations. Without initialisation, hashing,
-   * finalising or serialising this custom object will lead to crashes. See
-   * GPR#9513 for more details.
-   */
-  if (tg == Custom_tag)
-    caml_invalid_argument ("Obj.new_block");
-
   /* When [tg < No_scan_tag], [caml_alloc] returns an object whose fields are
    * initialised to [Val_unit]. Otherwise, the fields are uninitialised. We aim
-   * to avoid inconsistent states in other cases.
-   *
-   * For [Abstract_tag], [Double_tag] and [Double_array_tag], the initial
-   * content is irrelevant. [Custom_tag] objects are disallowed.
-   *
-   * For [String_tag], the initial contents do no matter. However, the length
-   * of the string is encoded using the last byte of the block. For this
-   * reason, the blocks with [String_tag] cannot be of size [0]. We initialise
-   * the last byte to [0] such that the length returned by [String.length] and
-   * [Bytes.length] is non-negative number.
-   */
+   * to avoid inconsistent states in other cases. */
   res = caml_alloc(sz, tg);
 
-  if (tg == String_tag) {
+  switch (tg) {
+  case Closure_tag: {
+    /* [Closure_tag] is below [no_scan_tag], but closures have more
+       structure with in particular a "closure information" that
+       indicates where the environment starts. We initialize this to
+       a sane value, as it may be accessed by runtime functions. */
+    /* Closinfo_val is the second field, so we need size at least 2 */
+    if (sz < 2) caml_invalid_argument ("Obj.new_block");
+    Closinfo_val(res) = Make_closinfo(0, 2);
+    break;
+  }
+  case Abstract_tag:
+  case Double_tag:
+  case Double_array_tag: {
+   /* For [Abstract_tag], [Double_tag] and [Double_array_tag], the initial
+      content is irrelevant. */
+    break;
+  }
+  case String_tag: {
+    /* For [String_tag], the initial content does not matter. However, the length
+       of the string is encoded using the last byte of the block. For this
+       reason, the blocks with [String_tag] cannot be of size [0]. We initialise
+       the last byte to [0] such that the length returned by [String.length] and
+       [Bytes.length] is a non-negative number. */
     if (sz == 0) caml_invalid_argument ("Obj.new_block");
     Field (res, sz - 1) = 0;
+    break;
+  }
+  case Custom_tag: {
+    /* It is difficult to correctly use custom objects allocated
+       through [Obj.new_block], so we disallow it here. The first
+       field of a custom object must contain a valid pointer to
+       a block of custom operations. Without initialisation, hashing,
+       finalising or serialising this custom object will lead to
+       crashes.  See #9513 for more details. */
+    caml_invalid_argument ("Obj.new_block");
+  }
```

Comment:
I'm sorry to be difficult, but the `Custom_tag` case must be rejected BEFORE doing `caml_alloc`.  The reason is as follows: if the block is big enough, it is allocated in the major heap, with tag `Custom_tag` but an invalid "operations" field.  Then, we raise an exception, so the block is unreachable.  So, at the next major GC sweep, it will be collected, and the sweeper will look into the "operations" field to see if there is a finalization function.  This will segfault.

Reply:
I see. I noticed that the allocation would happen, but I figured that this would be fine given that this exception corresponds to a programming error, that is not supposed to be handled by the program. Your point about the segfault is very reasonable :-)

For the other failure cases in the function, there is no such issue: there are failure cases that can lead to ill-formed values on the heap, but they are not ill-formed enough that it would be a problem for the GC, which is the only case we have to consider for this allocation that is dead on arrival.

Reply:
@xavierleroy I moved back the Custom_tag check, but then I pushed a second commit that moves the allocation within the switch, so that it is always after the failure test. Let me know what you prefer.